csds-client/client/client_test.go

```diff
@@ -177,6 +178,28 @@ func TestParseResponseWithNodeId(t *testing.T) {
 	}
 }
 
+// TestVisualization tests parsing xds relationship from config and generating .dot
+func TestVisualization(t *testing.T) {
+	filename, _ := filepath.Abs("./response_for_visualization.json")
+	responsejson, err := ioutil.ReadFile(filename)
+	if err != nil {
+		t.Errorf("Read From File Failure: %v", err)
+	}
+	graphData, err := parseXdsRelationship(responsejson)
+	if err != nil {
+		t.Errorf("Parse xDS relationship failure: %v", err)
+	}
+	dot, err := generateGraph(graphData)
+	if err != nil {
+		t.Errorf("Generate graph failure:%v", err)
+	}
+	want := "digraph G {\nrankdir=LR;\n\\\"test_lds_0\\\"->\\\"test_rds_0\\\"[ arrowsize=0.3, penwidth=0.3 ];\n\\\"test_lds_0\\\"->\\\"test_rds_1\\\"[ arrowsize=0.3, penwidth=0.3 ];\n\\\"test_rds_0\\\"->\\\"test_cds_0\\\"[ arrowsize=0.3, penwidth=0.3 ];\n\\\"test_rds_0\\\"->\\\"test_cds_1\\\"[ arrowsize=0.3, penwidth=0.3 ];\n\\\"test_rds_1\\\"->\\\"test_cds_1\\\"[ arrowsize=0.3, penwidth=0.3 ];\n\\\"test_cds_0\\\" [ color=\\\"#34A853\\\", fillcolor=\\\"#34A853\\\", fontcolor=white, fontname=Roboto, label=CDS0, shape=box, style=\\\"\"filled,rounded\"\\\" ];\n\\\"test_cds_1\\\" [ color=\\\"#34A853\\\", fillcolor=\\\"#34A853\\\", fontcolor=white, fontname=Roboto, label=CDS1, shape=box, style=\\\"\"filled,rounded\"\\\" ];\n\\\"test_lds_0\\\" [ color=\\\"#4285F4\\\", fillcolor=\\\"#4285F4\\\", fontcolor=white, fontname=Roboto, label=LDS0, shape=box, style=\\\"\"filled,rounded\"\\\" ];\n\\\"test_rds_0\\\" [ color=\\\"#FBBC04\\\", fillcolor=\\\"#FBBC04\\\", fontcolor=white, fontname=Roboto, label=RDS0, shape=box, style=\\\"\"filled,rounded\"\\\" ];\n\\\"test_rds_1\\\" [ color=\\\"#FBBC04\\\", fillcolor=\\\"#FBBC04\\\", fontcolor=white, fontname=Roboto, label=RDS1, shape=box, style=\\\"\"filled,rounded\"\\\" ];\n\n}\n"
```

Comment:
Have a comment to "manually" review this? A link to that viewer site with this data?

Reply:
I modified this test to open two windows in the browser to show the expected graph and the generated graph. Does this make sense?